notes/plutus-report/costing.tex

```diff
@@ -0,0 +1,339 @@
+
+\paragraph{Builtin accounting happens before execution}
+Builtin operations can in principle use a lot of resources just in the operation itself.
+For example, multiplying two sufficiently large numbers can require a large amount of memory to hold the result, much larger than that required to hold the arguments.
+This leads to a conflict: builtin operations are atomic (in that we can't interrupt them), but we want to interrupt evaluation as soon as we exceed our resource budget.
+
+Our solution is to compute the costs for a builtin operation before we run it.
+Since we are not actually measuring anything, but rather computing the cost from our model, we are able to do this.
+That way we can terminate evaluation if we would exceed the budget after running the operation, without actually running the operation.
+
+\paragraph{Time for builtin operations}
+The approach we take for Time is very empirical:
+\begin{enumerate}
+\item Micro-benchmark the builtin with arguments of varying sizes.
+\item Look at the resulting graphs and try to pick an appropriate linear model for the running time based on the argument sizes (e.g. ``linear in both arguments'').
+\item Use linear regression to infer the parameters for the model, these are the cost model parameters for this builtin.
+\end{enumerate}
+
+\paragraph{Space for builtin operations}
+The approach we take for Space is more deductive.
+Generally it is hard to measure memory allocation, especially when it occurs in a foreign library.
+So we largely rely on reasoning about or inspecting the libraries in question.
+Many perform no allocation, or a predictable amount of allocation.
+
+\subsubsection{Space and Time for machine steps}
+The CEK machine itself has many steps that it goes through, which correspond to handling e.g. application of user-defined functions.
+These also need to have costs associated: naive profiling suggests that the operation machine itself (as opposed to builtin functions) is usually >50\% of evaluation time.
+
+\paragraph{Time for machine steps}
+The approach we initially took for Time is empirical, matching the approach for builtins:
+\begin{enumerate}
+\item Assume that the execution time is linear in all the kinds of machine steps.
+\item Benchmark execution of a large number of varied programs that use all the execution steps.
+\item Use linear regression to infer the parameters for the model, these are the cost parameters for the machine steps.
+\end{enumerate}
+
+However, when we actually did this, we found that the number of machine steps of different kinds are almost perfectly correlated, such that there is little point trying to infer coefficients for them individually.
+So instead we fit a single-parameter linear model based only on the total number of steps.
+
+\paragraph{Space for machine steps}
+The approach we take for Space is more deductive.
```

Comment:
At the moment it's nonexistent!  I will think about that once we've got a reasonable time model though.  I think we'll just have to come up with something we think is reasonable and see how it works.  The difficulty of measuring the memory usage makes it even trickier.

Reply:
Yeah, this is somewhat theoretical and based on what I thought we were going to do.